src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/SymbolicRegexMatcher.cs

```diff
@@ -769,189 +607,282 @@ private int FindEndPositionCapturing(ReadOnlySpan<char> input, int i, out Regist
+                            {
+                                // no match was found
+                                return NoMatchExists;
+                            }
+
+                            initialStateIndex = i;
+
+                            // Update the starting state based on where TryFindNextStartingPosition moved us to.
+                            // As with the initial starting state, if it's a dead end, no match exists.
+                            currentState = new CurrentState(_dotstarredInitialStates[GetCharKind(input, i - 1)]);
+                            if (currentState.DfaState!.IsNothing)
+                            {
+                                return NoMatchExists;
+                            }
                         }
                     }
-                }
 
-                int result;
-                int j = Math.Min(input.Length, i + AntimirovThresholdLeeway);
-                bool done = _builder._antimirov ?
-                    FindFinalStatePositionDeltas<AntimirovTransition>(input, j, ref i, ref q, ref matchLength, out result) :
-                    FindFinalStatePositionDeltas<BrzozowskiTransition>(input, j, ref i, ref q, ref matchLength, out result);
+                    // Now run the DFA or NFA traversal from the current point using the current state.
+                    int result;
+                    bool done = currentState.NfaState is not null ?
+                        FindFinalStatePositionDeltas<NfaStateHandler>(builder, input, ref i, ref currentState, ref matchLength, out result) :
+                        FindFinalStatePositionDeltas<DfaStateHandler>(builder, input, ref i, ref currentState, ref matchLength, out result);
 
-                if (done)
-                {
-                    return result;
-                }
+                    // If we reached a final state, we're done.
+                    if (done)
+                    {
+                        return result;
+                    }
 
-                if (_checkTimeout)
-                {
-                    DoCheckTimeout(timeoutOccursAt);
+                    // If we didn't reach a final state and instead ran out of input, bail.
+                    if ((uint)i >= (uint)input.Length)
+                    {
+                        break;
+                    }
+
+                    // We didn't reach a final state, and we didn't run out of input, which means we either hit an initial
+                    // state (in which case FindFinalStatePositionDeltas will return in order to allow our initial state
+                    // optimizations to apply) or we tried to expand our node graph and exceeded our limits.  In the latter
+                    // case, we need to upgrade our DFA state to an NFA state such that processing continues in NFA mode.
+                    if (currentState.DfaState is not null && !currentState.DfaState.IsInitialState)
```

Comment:
Hmm, yes, if TryTakeTransition fails to create a new state because we've hit the limit _and_ we're in an initial state, this could be an infinite loop.  Got a pattern and input handy that should trigger that?  I guess I was assuming that wasn't possible, that if you were in an initial state you either wouldn't have exceeded your graph size or the node you'd need to transition to would have already been created, but that's a faulty assumption apparently.  I'll need to change FindFinalStatePositionDeltas then to be able to return three values instead of two, so that this code can determine whether TryTakeTransition failed and IsInitialState is true.  That shouldn't have a measurable impact on the inner loop.

Reply:
A pattern like `a{20000}|c` with an input like `aaaa...aaabc` where there's exacly enough a's to get to the edge (~10000) should do it as long as no prefix optimizations apply. Maybe the `c` part would need to be modified to fool the optimizations.

Reply:
I pushed a change to address this, but I was unable to come up with a repro.

We really need a more comprehensive set of tests for exercising NFA mode and the fall over into NFA mode.